pages/book/defining-types.mdx

```diff
@@ -0,0 +1,144 @@
+# Defining composite types
+
+import { Callout } from 'nextra/components'
+
+Tact supports a number of [primitive data types](/book/types#primitive-types) that are tailored for smart contract use. However, using individual means of storage often becomes cumbersome, so there are two main ways to combine multiple primitives together: [Structs](#structs) and [Messages](#messages).
+
+Note, while Traits and Contracts are also considered a part of the Tacts type system, one can't pass them around like [Structs](#structs) or [Messages](#messages). Instead, one can obtain the initial state of the given Contract by using the [initOf](/book/statements#initof) statement described later in the Book.
+
+<Callout type="warning" emoji="⚠️">
+
+  **Warning**: Currently circular types are **not** possible. This means that struct/message **A** can't have a field of a struct/message **B** that has a field of the struct/message **A**.
+
+  Therefore, the following code **won't** compile:
+
+  ```tact
+  struct A {
+      circularFieldA: B;
+  }
+
+  struct B {
+      impossibleFieldB: A;
+  }
+  ```
+
+</Callout>
+
+## Structs
+
+Structs can define complex data types that contain multiple fields of different types. They can also be nested.
+
+```tact
+struct Point {
+    x: Int as int64;
+    y: Int as int64;
+}
+
+struct Line {
+    start: Point;
+    end: Point;
+}
+```
+
+Structs can also include both default fields and optional fields. This can be quite useful when you have many fields but don't want to keep respecifying them.
+
+```tact
+struct Params {
+    name: String = "Satoshi";  // default value
+    age: Int?;                 // optional field
+    point: Point;              // nested Structs
+}
+```
+
+Structs are also useful as return values from getters or other internal functions. They effectively allow a single getter to return multiple return values.
+
+```tact
+contract StructsShowcase {
+    params: Params; // Struct as a Contract persistent state variable
+
+    init() {
+        self.params = Params{point: Point{x: 4, y: 2}};
+    }
+
+    get fun params(): Params {
+        return self.params;
+    }
+}
+```
+
+The order of fields does not matter. Unlike other languages, Tact does not have any padding between fields.
+
+## Messages
+
+Messages can hold [Structs](#structs) in them:
+
+```tact
+struct Point {
+    x: Int;
+    y: Int;
+}
+
+message Add {
+    point: Point; // holds a struct Point
+}
+```
+
+Messages are almost the same thing as [Structs](#structs) with the only difference that Messages have a 32-bit integer header in their serialization containing their unique numeric id. This allows Messages to be used with [receivers](/book/receive) since the contract can tell different types of messages apart based on this id.
```

Comment:
We should explain why anyone would want a "custom numeric id" referred to in `types.mdx`